{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Code for preprocessing alleleScan (ref) data \\\n",
    "Starting from the aligned pseudodiploid bams, we separate them out into wA 0,1,2 \\\n",
    "Then we get allele specific coverage and sequences accordingly \\\n",
    "We save this temporary result in a data_noGC_ref.h5 and then perform GC regression on it \\\n",
    "NOTE: This script only generates data_ident_ref.h5 (only ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# General import, names of train, test, val\n",
    "import numpy as np\n",
    "import pysam\n",
    "from tqdm.notebook import tqdm\n",
    "import h5py\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "rng = np.random.default_rng(seed=0)\n",
    "\n",
    "basedir = '/data/leslie/shared/ASA/'\n",
    "aligndir = f'{basedir}pseudodiploid/atac/'\n",
    "ctype = 'cd8'\n",
    "ident = '_vi_150bp'     # ident is '' for you or whatever label you want to use for the h5 data you've generated\n",
    "\n",
    "datadir = f'{basedir}mouseASA/{ctype}/data/'\n",
    "chrom_train = [1,2,4,6,7,8,9,11,12,13,14,16,17,18,19]\n",
    "chrom_val = [3,5]\n",
    "chrom_test = [10,15]\n",
    "chroms = [chrom_train, chrom_val, chrom_test]\n",
    "name = ['train', 'val', 'test']\n",
    "\n",
    "reps = ['r1','r2','r3','r4','r5']\n",
    "seqlen = 300                         # region around summit for sequence\n",
    "seqlen1 = 150                        # region around summit for coverage\n",
    "save = True                          # failsafe to prevent unwanted overwriting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocessing of Model Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "# some utility functions\n",
    "from utils import get_summits, get_shifts, one_hot, GCregress\n",
    "\n",
    "# make data directory for the cell type - done ONCE\n",
    "if not os.path.exists(datadir):\n",
    "    os.makedirs(datadir)\n",
    "    # separate out replicate BAM files based on allele tag wA, to be used for pileup purposes\n",
    "    for rep in reps:\n",
    "        bamfile = aligndir+ctype+'/'+rep+'.combined.rmDup.Aligned.sortedByCoord.out.bam'\n",
    "        _ = pysam.index(bamfile)\n",
    "        bamf = pysam.AlignmentFile(bamfile, \"rb\")\n",
    "        temp1 = pysam.AlignmentFile(datadir+rep+'_b6.bam', 'wb', template=bamf)\n",
    "        temp2 = pysam.AlignmentFile(datadir+rep+'_cast.bam', 'wb', template=bamf)\n",
    "        temp3 = pysam.AlignmentFile(datadir+rep+'_both.bam', 'wb', template=bamf)\n",
    "        i=0\n",
    "        for read in bamf.fetch():\n",
    "            try:\n",
    "                idx = read.get_tag('wA')\n",
    "                if idx==0:\n",
    "                    temp1.write(read)\n",
    "                elif idx==1:\n",
    "                    temp2.write(read)\n",
    "                elif idx==2:\n",
    "                    temp3.write(read)\n",
    "            except:\n",
    "                i+=1\n",
    "        print(i)\n",
    "        temp1.close()\n",
    "        temp2.close()\n",
    "        temp3.close()\n",
    "        bamf.close()\n",
    "\n",
    "        _ = pysam.index(datadir+rep+'_b6.bam')\n",
    "        _ = pysam.index(datadir+rep+'_cast.bam')\n",
    "        _ = pysam.index(datadir+rep+'_both.bam')\n",
    "\n",
    "N = []\n",
    "for rep in reps:\n",
    "    bamfile = aligndir+ctype+'/'+rep+'.combined.rmDup.Aligned.sortedByCoord.out.bam'\n",
    "    bamf = pysam.AlignmentFile(bamfile, \"rb\")\n",
    "    N.append( sum([bamf.get_index_statistics()[i][1] for i in range(len(chrom_train+chrom_test+chrom_val))]) )\n",
    "    bamf.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# process peakatlas file to get peaks dataframe\n",
    "if ident[:3]=='_vi':   # This is us using a peak atlas, you would be using the else clause here\n",
    "    summits = pd.read_csv(aligndir+'cd8_old/yi_cd8_peaks_33143.csv',sep=',',index_col=0)\n",
    "    summits['start'] += seqlen//2\n",
    "    summits = summits.iloc[:,1:3]\n",
    "    summits.columns = range(2)\n",
    "    summits_neg = pd.read_csv(aligndir+'cd8_old/yi_cd8_uneg_peaks_33143.csv',sep=',',index_col=0)\n",
    "    summits_neg['start'] += seqlen//2\n",
    "    summits_neg = summits_neg.iloc[:,1:3].reset_index(drop=True)\n",
    "    summits_neg.columns = range(2)\n",
    "else:\n",
    "    if ident=='_yi':\n",
    "        peakfile = aligndir+'cd8_old/peaks_yi.bed'\n",
    "    else:\n",
    "        peakfile = aligndir+ctype+'/peaks/peakatlas.bed'\n",
    "    peaks = pd.read_csv(peakfile, sep='\\t', header=None)\n",
    "\n",
    "    if not ident:\n",
    "        # filter peaks based on IDR threshold (column 12 of BED)\n",
    "        idr_thresh = 0.05\n",
    "        idx = np.where( peaks.iloc[:,11] >= -np.log10(idr_thresh) )[0]\n",
    "        peaks = peaks.loc[idx].reset_index(drop=True)\n",
    "\n",
    "    ## remove blacklist regions - try with and without\n",
    "    blacklistfile = aligndir+'mm10-blacklist.v2.bed'\n",
    "    blacklist = pd.read_csv(blacklistfile, sep='\\t', header=None)\n",
    "    peaks = remove_blacklist(peaks, blacklist)\n",
    "\n",
    "    # get summits of called + flanking neg peaks (moving away from this to background unegs)\n",
    "    summits, summits_neg = get_summits(peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Code to shortlist negative flanking summits to low accessibility - not needed if you are doing background unegs\n",
    "# Shit code, haven't used in forever, might need overhauling\n",
    "if ident[:3]!='_vi':\n",
    "    # FOR UNEGS, get the shortlist of summits_neg that have low accessibility\n",
    "    # get indices for each rep\n",
    "    idx = dict()\n",
    "    for rep in tqdm(reps):\n",
    "        bamfile = aligndir+ctype+'/'+rep+'.combined.rmDup.Aligned.sortedByCoord.out.bam'\n",
    "        bamf = pysam.AlignmentFile(bamfile, \"rb\")\n",
    "\n",
    "        idx[rep] = []\n",
    "        for chrom in chroms[:2]:\n",
    "            for c in chrom:\n",
    "                chromsummits = summits_neg.loc[np.where(summits_neg[0]==c)[0]]\n",
    "                y_uneg = []\n",
    "                for i in chromsummits[1]:\n",
    "                    y_uneg.append(bamf.count(str(c),i-seqlen//2,i+seqlen//2))\n",
    "                idx[rep] += list(chromsummits.index[ np.where(np.array(y_uneg)<5)[0] ])\n",
    "        bamf.close()\n",
    "\n",
    "    del y_uneg\n",
    "    # get low accessibility indices across reps\n",
    "    temp = set(idx[reps[0]])\n",
    "    for rep in reps[1:]:\n",
    "        temp.intersection_update(idx[rep])\n",
    "    idx = list(temp)\n",
    "\n",
    "    # shortlist unegs to low accessibility indexes\n",
    "    summits_neg = summits_neg.loc[idx].reset_index(drop=True)\n",
    "    frac=0.5\n",
    "    if len(summits_neg) > frac*len(summits):\n",
    "        idx = rng.choice(len(summits_neg), int(frac*len(summits)), replace=False)\n",
    "        summits_neg = summits_neg.loc[idx].reset_index(drop=True)\n",
    "    summits_neg = summits_neg.sort_values(by=[0,1], ignore_index=True)\n",
    "\n",
    "    if save:\n",
    "        summits.to_csv(datadir+'summits'+ident+'.csv', index=False, header=False)\n",
    "        summits_neg.to_csv(datadir+'summits_neg'+ident+'.csv', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ef0eb3ed79294a8bbedac43586e12b96",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Generate y\n",
    "# Use summits and BAM files to generate pileup summed accessibilities\n",
    "y = dict()\n",
    "for rep in tqdm(reps):\n",
    "    temp1 = pysam.AlignmentFile(datadir+rep+'_b6.bam', 'rb')\n",
    "    temp2 = pysam.AlignmentFile(datadir+rep+'_cast.bam', 'rb')\n",
    "    temp3 = pysam.AlignmentFile(datadir+rep+'_both.bam', 'rb')\n",
    "\n",
    "    for n,chrom in zip(name,chroms):\n",
    "        y[rep+'_'+n+'_b6'] = []\n",
    "        y[rep+'_'+n+'_cast'] = []\n",
    "\n",
    "        for c in chrom:        # for each chromosome in train/val/test set\n",
    "            chromsummits = summits.iloc[np.where(summits[0]==c)[0],1]     # slice out the relevant chromosome summits\n",
    "            y_b6 = []\n",
    "            y_cast = []\n",
    "            for i in chromsummits:\n",
    "                # for each summit peak, get the summed accessibility from BAM pileup\n",
    "                y1 = temp1.count(str(c),i-seqlen1//2,i+seqlen1//2)\n",
    "                y2 = temp2.count(str(c),i-seqlen1//2,i+seqlen1//2)\n",
    "                y3 = temp3.count(str(c),i-seqlen1//2,i+seqlen1//2)\n",
    "                y_b6.append(y1+y3/2)\n",
    "                y_cast.append(y2+y3/2)            \n",
    "            \n",
    "            y[rep+'_'+n+'_b6'] += y_b6\n",
    "            y[rep+'_'+n+'_cast'] += y_cast\n",
    "    \n",
    "    temp1.close()\n",
    "    temp2.close()\n",
    "    temp3.close()\n",
    "    \n",
    "del y_b6, y_cast, y1, y2, y3, chromsummits\n",
    "\n",
    "# merge RPMs across reps\n",
    "for n in name:\n",
    "    y[n+'_b6'] = np.zeros_like(y[rep+'_'+n+'_b6'])\n",
    "    y[n+'_cast'] = np.zeros_like(y[rep+'_'+n+'_cast'])\n",
    "    for i,rep in enumerate(reps):\n",
    "        y[n+'_b6'] += np.array(y[rep+'_'+n+'_b6'])*1e6/N[i]     # get summed RPMs\n",
    "        y[n+'_cast'] += np.array(y[rep+'_'+n+'_cast'])*1e6/N[i] # get summed RPMs\n",
    "        del y[rep+'_'+n+'_b6'], y[rep+'_'+n+'_cast']\n",
    "\n",
    "    y[n] = np.log2( 1+y[n+'_b6']+y[n+'_cast'] )      # normalize summed counts to RPM\n",
    "    del y[n+'_b6'], y[n+'_cast']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1f0ea740e8874c7b93f2445854f7b3e9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/15 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "910c7dabdafc4ca9a1eacbe2655d150c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "33ab38df6b3e495b88942e473066e977",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Generate X\n",
    "from Bio import SeqIO\n",
    "from bisect import bisect\n",
    "\n",
    "gen_b6 = SeqIO.index(aligndir+'gen/b6.fa', 'fasta')\n",
    "\n",
    "x = dict()\n",
    "for n,chrom in zip(name,chroms):\n",
    "    x[n] = []\n",
    "    if n!='test' and summits_neg:\n",
    "        x[n+'_unegs'] = []\n",
    "\n",
    "    for c in tqdm(chrom):        # for each chromosome in train/val/test set\n",
    "        seq_b6 = ''.join(gen_b6.get_raw(str(c)).decode().split('\\n')[1:])\n",
    "\n",
    "        chromsummits = summits.iloc[np.where(summits[0]==c)[0],1]     # slice out the relevant chromosome summits\n",
    "\n",
    "        # get relevant b6 genomic seqs\n",
    "        x[n] += [seq_b6[i-seqlen//2:i+seqlen//2] for i in chromsummits]\n",
    "\n",
    "        if n!='test' and summits_neg:\n",
    "            chromsummits = summits_neg.loc[np.where(summits_neg[0]==c)[0],1]     # slice out the relevant chromosome summits\n",
    "            x[n+'_unegs'] += [seq_b6[i-seqlen//2:i+seqlen//2] for i in chromsummits]\n",
    "\n",
    "    x[n] = one_hot(x[n])             # convert string of nucleotides to one-hot representation\n",
    "    if n!='test' and summits_neg:\n",
    "        x[n+'_unegs'] = one_hot(x[n+'_unegs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# new code for actual background unegs - can be merged into x code later - SKIP till SKIP END if dont need to compute\n",
    "# written very stupidly, basically just overwrite the old unegs\n",
    "# this is because I didn't want to write logic to do one or the other depending on which method of unegs you want\n",
    "from utils import get_neg_summits\n",
    "\n",
    "bamfile = aligndir+ctype+'/'+reps[0]+'.combined.rmDup.Aligned.sortedByCoord.out.bam'\n",
    "bamf = pysam.AlignmentFile(bamfile, \"rb\")\n",
    "\n",
    "for n,chrom in zip(name,chroms[:2]):\n",
    "    x[n+'_unegs'] = []\n",
    "    \n",
    "    for c in tqdm(chrom):        # for each chromosome in train/val/test set\n",
    "        seq_b6 = ''.join(gen_b6.get_raw(str(c)).decode().split('\\n')[1:])\n",
    "        \n",
    "        chromsummits = summits.iloc[np.where(summits[0]==c)[0],1]     # slice out the relevant chromosome summits\n",
    "        neg_summits = np.empty(0, dtype=np.int64)\n",
    "        while len(neg_summits)<len(chromsummits):     # get neg summits and only keep low coverage ones (<5)\n",
    "            temp = get_neg_summits(chromsummits, len(chromsummits)-len(neg_summits), len(seq_b6))\n",
    "            idx = np.where(np.array([bamf.count(str(c),i-seqlen//2,i+seqlen//2) for i in temp]) < 5)[0]\n",
    "            temp = temp[idx]\n",
    "            neg_summits = np.concatenate((neg_summits, temp))\n",
    "        neg_summits = np.sort(neg_summits)\n",
    "        \n",
    "        # get relevant b6 genomic seqs\n",
    "        x[n+'_unegs'] += [seq_b6[i-seqlen//2:i+seqlen//2] for i in neg_summits]\n",
    "        \n",
    "    x[n+'_unegs'] = one_hot(x[n+'_b6_unegs'])\n",
    "# SKIP END - comment upto here if you want to skip the background sampling and go with flank regions instead\n",
    "\n",
    "gen_b6.close()\n",
    "gen_cast.close()\n",
    "\n",
    "# set uneg log acc to low constant value (at tail end of post-GC residual accessibility distribution)\n",
    "for n,chrom in zip(name[:2],chroms[:2]):\n",
    "    y[n+'_unegs'] = -5.0*np.ones(len(x[n+'_unegs']))\n",
    "\n",
    "# Finally, save everything...\n",
    "if save:\n",
    "    with h5py.File(datadir+'data_noGC'+ident+'_ref.h5','w') as f:\n",
    "        for key in x.keys():\n",
    "            f.create_dataset('x_'+key, data=x[key])\n",
    "        for key in y.keys():\n",
    "            f.create_dataset('y_'+key, data=y[key])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform GC regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = dict()\n",
    "y = dict()\n",
    "with h5py.File(datadir+'data_noGC'+ident+'_ref.h5','r') as f:\n",
    "    for n in name:\n",
    "        x[n] = f['x_'+n][()]\n",
    "        y[n] = f['y_'+n][()]\n",
    "    x['train_unegs'] = f['x_train_unegs'][()]\n",
    "    x['val_unegs'] = f['x_val_unegs'][()]\n",
    "    y['train_unegs'] = f['y_train_unegs'][()]\n",
    "    y['val_unegs'] = f['y_val_unegs'][()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlYAAAEvCAYAAACHYI+LAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO3df6zd9X3f8dfb10n44ZF0xtAVQuwyp1uG4ia6qWkjlSpNIcOZyR+bxChRlk5Clbo2Y4taI0SRFUW11Gqs0rZKKFlTxSEsYhlBM6uddhvT/sDjktSLU1qcxIbgpHDDVBiEDHz57I97bO65fn/vfX/v5/P9eZ4PCR04nHu+33O+55zv+/v+vD/vj6WUBAAAgHybut4BAACAsSCwAgAAKITACgAAoBACKwAAgEIIrAAAAAohsAIAAChkcxcbvfTSS9P27du72DQAAEAtjz/++A9SStsij+0ksNq+fbsWFha62DQAAEAtZvZU9LEMBQIAABRCYAUAAFAIgRUAAEAhBFYAAACFEFgBAAAUQmAFAABQCIEVAABAIQRWAAAAhRBYAQAAFNJJ53XMhuvveURPPvuS3nn5Fh25/bqudwcAgMaRsUJjnnz2palbAADGjsAKjXnn5VumbgEAGDuGAtEYhv8AALOGjBUAAEAhBFYAAACFEFgBAAAUQmAFAABQCIEVAABAIQRWAAAAhRBYAQAAFEJgBQAAUAiBFQAAQCEEVgAAAIUQWAEAABRCYAUAAFAIgRUAAEAhBFYAAACFEFgBAAAUUiSwMrPbzeybZnbczL5oZheUeF4AAIAhyQ6szOwKSb8haT6ldI2kOUk35z4vAADA0JQaCtws6UIz2yzpIknfK/S8AAAAg5EdWKWUTkv6PUlPS/q+pBdSSkdWP87MbjOzBTNbWFxczN0sAABA75QYCvwxSTdJ2iHpJyRdbGa3rn5cSunelNJ8Sml+27ZtuZsFAADonRJDgR+UdDKltJhSek3SlyX9XIHnBQAAGJQSgdXTkq41s4vMzCT9oqQnCjwvAADAoJSosToq6QFJX5P0jclz3pv7vAAAAEOzucSTpJTulnR3iecCAAAYKjqvAwAAFEJgBQAAUAiBFQAAQCEEVgAAAIUQWAEAABRCYAUAAFAIgRUAAEAhBFYAAACFEFgBAAAUQmAFAABQCIEVAABAIQRWAAAAhRBYAQAAFEJgBQAAUAiBFQAAQCEEVgAAAIUQWAEAABRCYAUAAFAIgRUAAEAhBFYAAACFEFgBAAAUQmAFAABQCIEVAABAIQRWAAAAhRBYAQAAFEJgBQAAUEiRwMrM3mZmD5jZX5jZE2b2syWeFwAAYEg2F3qe35f0xymlf2hmb5Z0UaHnBQAAGIzswMrMLpH085L+iSSllF6V9Gru8wIAAAxNiaHAn5S0KOkPzezrZvYZM7u4wPMCAAAMSonAarOk90r6g5TSeyS9LGnf6geZ2W1mtmBmC4uLiwU2CwAA0C8lAqtnJD2TUjo6+e8HtBxoTUkp3ZtSmk8pzW/btq3AZgEAAPolO7BKKf2VpO+a2U9N7vpFSX+e+7wAAABDU2pW4K9L+sJkRuB3JH280PMCAAAMRpHAKqX0Z5LmSzwXAADAUNF5HQAAoBACKwAAgEIIrAAAAAohsAIAACiEwAoAAKAQAisAAIBCCKwAAAAKIbACAAAohMAKGLG7Hjyuq+94WHc9eLzrXQGAmUBgBYzYfUef1lJKuu/o013vCgDMBAIrYMRu2X2V5sx0y+6rut4VAJgJllJqfaPz8/NpYWGh9e0CAADUZWaPp5RCayKTsQIAACiEwAoAAKAQAisAAIBCCKwAAAAKIbACAAAohMAKAACgEAIrAACAQgisAAAACiGwAgAAKITACgAAoBACKwAAgEI2d70DAID+uf6eR/Tksy/pnZdv0ZHbr+t6d4DBIGMFADjPk8++NHULIIaMFQDgPO+8fMu5jFUfkEHDUBQLrMxsTtKCpNMppQ+Xel4AG3fXg8d139Gndcvuq/Spj1zT9e5gA7oKKPoWvJBBw1CUHAr8hKQnCj4fgEz3HX1aSynpvqNPd70r2CACimVnM2d9yaABVYpkrMzsSkl7JH1a0r8o8ZwA8t2y+6pzGSsMU9+G5LrStwwaUMVSSvlPYvaApN+R9DckfXK9ocD5+fm0sLCQvV2gLWOq72B4EADqMbPHU0rzkcdmDwWa2YclPZdSenydx91mZgtmtrC4uJi7WaBVYxqOYXgQAJpTosbq/ZL2mtkpSfdL+oCZHVz9oJTSvSml+ZTS/LZt2wpsFmjPmOo7btl9lebMGB4EgAYUGQo892RmvyCGAtGiMQ3RdYnhQQCo1upQINAlb4ju+nse0fZ9h3T9PY90tVuDw/Dg2vhMAYgqGlillP47PazQJm+Irqoe6q4Hj+vqOx7WXQ8eb28HB4LhwbWNqcYOQLPovI5B84b/qqanr8zKzPJwlzfs96mPXDPT78l6aHkAIIrACkX0qUanqtZqywVzeuGVM9pywVzLe9QvBJj1Ub8HIIoaKxQxhBqdF145M3Xblr4NQTLsBwDNIbBCEZysqx189CktpaSDjz7V9a5IWh72+/bv3DiKbFXfglYAILDCzPjote/QnJk+eu07Wt1uWnVblxc8RAOKsQceQ8iUApgtBFYoYggnuJxMjRegeFPwd975sLbvO6Sddz68oX30tuO9t9593t9+fpIt+3xPsmWlkSkF0DcEVihiCCe4nOyNF8h4U/BfW0pTt1K9ru3esOGmTdO3knT1ZRdP3UrjD6I8YxrWrDL2rCMwNgRWI9fWj/IQTnA5gYcXOL71ws1Tt5L0pjmbupWk3Tu2as5Mu3dsPXdfVcNJb9jQC9a+/dzLU7e5mmiAOaaAoMvXMoRsMIA3EFh1JKdupo5o4XRb+xOVs+2qv7VVt1WP9e7zAkdvluHN71sOwG5+3xsBmBfQVTWc9PbRC9a8jFWUF0TVaYAZfc/GFBB0+VqGkA0G8AYCq45E62ZyRQun29qfqJzsUtV+75wMxe1cMSTnPTZn27nDcbdOCuxvXVFg//rr07eSdGISAJ3YQCdwL4iKBp2S/xq9AN4LCKJBWd+WkOkyuBlCNhj1jSmji2kEVh3xfqib+PGOzoTL3Z+cH4noSdR7nLfdqmyON3yW857XqZ2KeujYaS2lpIeOnT53n/d6cmcaruY9X1WQ6AVh3t97AUE0kPWCvy6zqgQ3KK1vbVhQDp3XO9LWEiLR7XiPq7OPBx99Smlye/Zvot3Yo8NQ3uO8LuJVz3fL7qvO7U/d13j9PY+cW9LkbBdurxu3aTm4sPP+T4w3vOi9nrdeuFkvvHJmqr7L86Y502tLaWoY0RN9PskPorwlX7zjn9P9/mzg9fkVnzHvviZ4x78tfVrVAOWUvjhCf5Cx6pEh16REMx6lMwxnT9CRE3U061BnYefVcn8svXoqjxeAeZmkE5++UacO7NGJT9+45vPt3XWF5sy0d9cVtfdZ8t8f7/iX7n7vveYmdLkI85B/F1Ctq756aB6BVY8MoUi1KjCK/kjk1CB5J9Emlqk5+YOXp26r7Nh3SNv3HdKOfYfW3EdP1ePOTGb+nVmqH5p5NWTRYdYhtGrwgk6vJk2Sdu0/rO37DmnX/sPn7osOJXr3NTHsG9W33wVqg8pgeHm8CKx6ZAhftC6vnptInXsnCa+9QbSuKLqPVY+L/r3X6sHLqrSVafH2pzTvuFR9Z7yAO1p07z3uyO3X6dSBPZ0sxty33wUyaMDaCKxQS1VheFcZjzrpdC/DFN3vSyYBwyUNBg51dLWgdJWu9id39qAXyLY1vDjUzE/fMmhA3xBYjVzpH+/SjSmleKNNjzeLrkpOxqtvgQyWNZGR84LonJmrVYaa+elbBg3oGwKrkcuZ0usFZXWKxT1eEOUFLd6wj5edIuAZljaGDHNFZ2ZGG61WBVte9neoWSwAbyCwGrmcLI13RV0VyHjDJ6VnvRw9+byWUtLRk88Xeb5Z1dZQl2eogbD3nuW0CZH87G/OkPoQgrIh7COQi8Bq5HKCmzq1FNEALufE2uWU9zGhf059Tbxn3uLaOUHvEIKyoQ5/AnUQWHVkbB2jveno/IgC1bzhbo83lBit5Yr+znjf1aq/zfntovAds4DAqiN1go7SQVj0R9nbxzr9fPgRBerxMmM5bTSiAZP3Xa36jcq5YKLwHbOAwKoFddaz85ReUyr6o+ztY9WsQIbpgP7xvsPekKEX8FRNVIkurg3MKgKrFnhXeHUCka5qYnIXLR5CN29gzHIueOrUQ0a/610upA20hcCqBblDYl4Beukfo+jwwP2PLQeJ9z9G3RQwFjvvfFjb9x3SzjsfXvexOUOB3t9Si4mxIbBqQW5dgff3dYpNI7zn87YbLbgFMBx1vtd1yhhW8y7Wcp4P6KPswMrM3m5m/83MnjCzb5rZJ0rs2Njl9hKqU2wa4dVTeAvZApgNVRNVcoYXvYu1JlZzALpUImN1RtK/TCn9XUnXSvo1M3tXgecdtVsnw3u3FmqeKcU7OXs/mF49RRPNHLtsTgkgrq0JKd7vVk6LCaBr2YFVSun7KaWvTf79/0p6QtIVuc87drnDg1526sTkB/DEih9C73G7d2zVnJl279i65jaaCIJoTgkMW87vQnQZoJwWE0DXitZYmdl2Se+RdLTk8+J83lWeF7REp1t72gqChrB+HIBl0d8FL1seDY683wR+JzAUxQIrM9si6T9K+ucppRed/3+bmS2Y2cLi4mKpzY5KneJz7wfKmz0YrV9o4kcremU71PXjAFTLabfi/Sa89KOlqVugr4oEVmb2Ji0HVV9IKX3Ze0xK6d6U0nxKaX7btm0lNjs6dRqBRoOWaKuHJoIbhv0AlOL9llXVXVGPhS6VmBVokj4r6YmU0r/K36XZVScQ8Yrfoy0TSKkDGJqHjp3WUkp66Njpc/dVDS1GhxxpToomlMhYvV/SRyV9wMz+bPLPjQWed5CiX1Tvcd5QXpU6S1Cs5mWn3jRnU7dSvXUBAaBJuVl1L4tFc1I0ocSswP+ZUrKU0rtTSj89+Wf99r0jFf2ieo87evJ5LaWkoyefX3c7XmDm/fBEA70zk8aAZ1Y0CPRmD9IuAcAQeVksmpOiCXReLyxa0+R9oetMJ44WhnqP8zJR3jCk97eXTIYPL2EYEcCAeBeFNCdFEzg7FhbNOnlf6HdevkVPPvtSaOjNtBwEbSRztHvHVn37uZfX7WPlYQYfgCHyLh43bZKWlpZvz9q1/7BeeOWM3nrhZh27+4Y2dxEjMfMZq9KzR6JZJy+zVdW40xvOy+ncTl0BAPhrJHoXjzv2HdL2fYe0Y9+hNZ+vzmLWGK+ZCayqao1Kd/ONFnx7xedV7RaigZBXgO6JFrkDAOIztlmkHtIMBVZVwUnp1gNHbr9Opw7s0ZHbr1vzcV6gV/Xl9eqxvNfjFaB7GM4DgDy5ox20ehivmQmsqorKu+rm6wVGVe0WvHos7/V4gVnpWXx1WkIAwFjl9sqKNoQmABuemQmsqhY9js7iK83bbp199IrkvSBq82RYcPM6w4MAgPKqRkuiw4vUxA7PzARWVaqCmYicK4k62/WCKO9qyStoLz3mn7P+FwDMmjo1rd45pauLf2zczAdWOcFRW1cS0ZRzTpAIACivTk1rdFmyHAwtNm/mA6uc4KjLrr3R2Yd0SgeAYfCyW6UDIYYWmzfzgVVOmjXatTf3i+EVjHtZLG87OTVWBGUA0B4vu+WVX+ScU7xzHlmssmY+sMpJs0aDsiauELygx/sCejVW0X5X0eJKAEC+6G9zzjnFO+eRxSprZgKrqoi8jQL03CsEL2CKdl73ArDXX5++BQB0z7sQ9n7DSw8ZUiBf1igDK+8DVhWRtxGp514h5AzJeVmnaG1Y9OoJANAM7zc8OmQYxcSnskYZWHlBS1VEHo3US49Be8FN1Ta87JT3GqMF7ScmdVkn1pllyPIMAADUU2Ydl565ZfdVuu/o0+c13/Si8ar7V1sZyJSI6r3C97NXGp9/9KmpbXj7uOWCOb3wypmpVHC0LQO1UwAwDG+9cLNeeOXM1LJrpuXf7/XGEnbtP3zub4/dfUOTu4kVRpmxakLOGLSXicpt1cB6fwAwft6ya9GLY84T3RhlYNVE3VTOGLS3P9FWDbmiw4MAgP7psl+ih9YM6xtlYNW3GQ7e/nhflqogyPsgRwOmtgI4AEB50RIPj9cD0Tuf1Jk1z7Jm6xtlYNXEDIfSbRm8gGf3jq2aM9PuHVun/t77IB+5/TqdOrBHR26/bs1t9y3IBAC0IzojvctZ82M0ysCqCTkfsGiNVZ1tRAM9ptECAM7yzj1VC0V797Mix/oIrIJyMj9exslreVBnG9F2C4yHA8Bsuv6eR7R93yFdf88j5+7zRkuqity9+zeZTd3ifKMMrJoIJrzMT852vFkdVdmls9NsV0639a46vKHEg5Og7iDj4QAwCtEaW68+6/WUpm7X4jWJprxkfaMMrNoaF/a2Ew22vHSqd3Uh+dNtq/pgrc6Mle5Z5QV5AID2eAFT9Le5zjkh2iTaO+9Vnc9mwSgDq7Yiam87XrDlBVFeN/Wq2R/edqKvsXQgRF8UAOifH766NHVbhzd7sIp3jvMu6nNmMw7dKAOrtgq2ve14AY8XROXuY/TvvUCINQABYFyiCzh7Hjp2Wksp6aFjp9fdTvSiPjpcOcY64CKBlZl9yMz+0sy+ZWb7Sjxn35Q++N7zVV01eFcIOWlW1gAEgPFrokO7d1HvBVHRlkBj7IuVHViZ2Zykfyvp70t6l6R/bGbvyn3evonWbUV7hHj3VWWhvEJ1L81Kl3UAQNu8ICqn3njoSmSsfkbSt1JK30kpvSrpfkk3FXjeXommP6P1UHXqwLxCda92KnqFAACYTXXqqTzR0ZJoMsIrlRm6EoHVFZK+u+K/n5ncNyo5NVHe39Z5Pi8I81K3YxyrBgCUU6eeyuONlnjnnlluy1AisPIyeOcN6ZrZbWa2YGYLi4uLBTbbT030jfKCMC99Gm0aCgCYTU3M7K5T2rIaNVa+ZyS9fcV/Xynpe6sflFK6N6U0n1Ka37ZtW4HN9lPpvlGSfzWwcxIo7VwRMHlXCF7TUGYFAgBKqVoSZzVvGJEaK99jknaa2Q4ze7OkmyU9VOB5i2prmCx3/NrjXQ14dVce72rg5vctB2A3v2/2UrQAgI3zLsy9LNiu/Ye1fd8h7dp/+Nx93jAiNVaOlNIZSf9M0mFJT0j6Ukrpm7nPW1pb3dib6KEVLX6PplQZMgQAbITXrsc7f3jBljfpqq2+k20q0o47pfSwpIdLPFdTbtl9le47+vR5hXR3PXj83P19PbCf+sg1RfdtywVzeuGVM+umbQEAWE90Jrq3PNsYjbLzuqcqKm4rk5XDG8bMKfjzriRmefkBAEDzZmWm4MwEVlWGcKBzgj8vRUvxOgDMptLrx0rxGuYxDvt5yr2zA1V6mK0JV192sZ589qWpzuse0/JsxJXhkpeiZUkbAOjWm+ZMry2lRi9w33rhZr3wypmiQZRn5cX/2fPpOy/foieffWnqon4IpTclzHxgNQQnJsNzJ9YZprv12ne4dWQAgH5p4wLXK/tooo/Vpk3S0tLy7VneRf3Z0pXPP/rUqAOrmR8KHIKc3lheipahQACYTU38/ntB4iyvBEJgVVh0HaU6vAZq0c7r3n0MBQLAbPJ+/5uou2piFZKhILAqLGd2XVWE7zVQ8+47W4O1shYrWpzvXcXQ2woAhiva1Xzvris0Z6a9u8ot8+uNtEQbaNfJdvUxM0ZgVVhOMFI1+y86kyLajd3jXcUcuf06nTqwJ9yjBADQH9EykjrZpeg5zguivHOZFxjVmQnfx8wYgVVhOcFIVXbJG170+lh5GSvvcd4H3ksFNzGsCQDol6oAzAuivAt473FHTz6vpZR09OTza27bC6LqtEFqYn3eXARWPVKVmYoOL3ofeC8V7G2ny6ahY1yEEwCGoioL9eyLP5q6lfygx3tc9PzhPV80syU1sz5vLgKrAfA+9N593gc0Z4HLtmqs+njFAQBDkvN7XVVGEm3NkNPCIVrqklsq0yb6WA3A7h1b9e3nXtbuHVvP3ecNNXrNTr37rr/nkXON26ifAtAWr4kx3uA11YzK+S2vWj/WazDq9aJqoxFp1Xq/fUTGagCixXnR2RFeitYbjmP9QAAlkZ1em9cMOtoKIWd2XFXGKTpb0Htc6RGPPmamqhBYDUD0xyhnTcGdkw//znW+BLRgAIC4Or+Z3m99dJjNm6jkXTDX2Z/oOcXbds5Erj62UKiDwKpHcovzvFmBHq9nlXel5G03p6UDAMyaOgFG6a7oXqDmjURUZcXqzM4ryQvUhoTAqkfqZJy8IMwLerzHeT2rvC/gQ8dOayklPXTs9Ln7uvqiAcAQ7dp/WNv3HdKu/YfXfWzOqhhecBSdcX3s7ht06sAeHbv7hqn7c4bfhp51ykFg1SNVQYsXcHl1V97fR4M1LzvlpaDrfNFYkxDArGti0ePodjaZTd3W5QWF3rnCG17MadzZxBI7bRrmXo+UN4NP8mdDROuuojMpvG2/ac702lKaCoyqZhR+9Np3nLcd7+qLWUFAe7zvMGZH7ky6qovr1ecKb5gzZ6LC6szZ0BBYDZQXoKzMTp394HtfgujUWC8wqpopWBUUrsasIKA9LLheXzQY9S4m+yb6u1ynBU/0sd77c9eDx8/dN4TZfRvFUOAAeMN5XuNPbyjQG+duKzUNoH+aGGYZ+tDNStFg1KtBzRUtn/CG3qL1VN5QXtUFc04bHq9sZOhF6VEEVoU1UbAXbfmf80GO7nfVDMVZLlQEhqSJC6tZvFhr4jVHgzpvolJ0NKBOnaz3nF7w5/3+e2vNzsryZQRWheX0kqpSujGaFxxF9/v+x5Yfd/9j04+blSsRYOiYVNIP0X5SXjDitdbxsoZ1ZiRGt+0Ff975w8ts5SyxNiTDz9v2zBDa7nvj7ldfdrGefPaldXtgUbMBDBvf4X6I9gS81alV8oIWr+A7mlWrWkqnKmO1ugbNO+95zxmt+Ro6AqvCuvzgeIWB0QJC74vKjCIAaEb0YjYqZw3YOo/3AvPoTMFZwVBgYV3WGnnp2KMnn9dSSjp68vk1H+c58ekbderAHp349I2N7TMAzKLouoDRMouctV2rzlu5S+LMKjJWhXktD9ripWO9L5t3peS1YNi1//C5+4beVwRAtWgLllnlZe+9+6qG1DzeMFvp39noca06b3nDkLOciYrK+haZ2e9K+geSXpX0bUkfTyn9dYkdG6oua6yiw5BesOWNxXv38QMMIKJOkNF3F715Ti+8ckYXvXnu3H3ekFgTa6nm9NXyAjXvgrnqvOWdU2alF1WO3KHAr0q6JqX0bklPSrojf5eGrfQMvr6ZxWnVwNg18b0++YOXp25ztTWb0Rv+ir4/mzZN39blDcm9/vr0ba7cpcqipSSz3IInK7BKKR1JKZ09Oo9KujJ/lxCR86H1xsijDf5mpQ8J0BdD/c6Vnn3Y1mzGnNUhcvfRC1qiwVq0Fsv7ra86n3j3e60ecvZnjEqO5/yKpP9Q8Pmwhmgtl5ce9sbIo2P73pg7gOZcMhl+v2TFiTC65uaszuzt6nXnlkq8ntLUreQHa9HteDMFvd/6g48+pTS5Xd1g+uzt2fujw52zvC7suhkrM/sTMzvu/HPTisfcKemMpC+s8Ty3mdmCmS0sLi6W2fsZ5nVj90RTvF6XXADtig5DnQ2yLtnAep+z0CC0q15dl19ywdRtXdFs2YuTz8KL6wxNevW03m991Xa9z2N06bRoM9AxnnvWDaxSSh9MKV3j/PMVSTKzj0n6sKRfTilVfh5SSvemlOZTSvPbtm0r9wpmVOlaLu8LWKdD+5jWCgNWa2I4zvvORE+sOTVRN79v+cR48/vWviirmlbf1Xd9CAFhTsuDKjmfE+8Y1tlHLzjyzj3RVj+eJt6zrmXVWJnZhyT9lqS9KaUfltkldMH7AnpfoKpM2Us/Wpq6lYZbGwKsFs0Q1ZGb3dio3NqX0oXu0UAtNwtVuv9S1bqpObz3Ipqd8uzesVVzZtq9Y+ua26gSvYD36q6iAdMY+2Ll/kr8G0lvkfRVM5OkR1NKv5q9V2hdtDdJVUuHLRcsT0fecsEb05FzikCBPmli1lzOlXq0lUFOncuQMwleDVLfXo93bLzPWc7vqFcj5V0E5/LqrqJ1YGPsi5U7K/Bvp5TenlL66ck/BFUzKnriIYuFIerb5zYaJLS16K2XdYhmdKK/HXUyLaWDh2hH9DrZl2gW1Pvsee+tt+1ojVQu7zmbCOCGgiVtUET0h5UsFoaoy89tzpCTN5TTxNCLF+gdnAQeBwtNt69zos7pJ+UFUXt3XaE5M+3ddcWaf1snK+YFlF4g5AXHXv2SlzWK1kg1oYkAbigIrFDEkduv06kDe6bSut4XeAgFqECf5JwIvdla3gm4TvDmPdYLCHIKrD3eiboqk5hTj+UFPNG6tDqZTe+30HvP7n9suTD8/sfeKAz3Ajjv/fE+O95nouoYRPslesXrY2+WvRYCK7TKW9g5+qM+xiJHDFdbn8ecZsBe1ija4LGKd8LcOXkPdm7gvYhmebzt9i0DXud9iAZ/pVtHeEGid2Fc9VjPLGenPARW6Jz3g+kFW1VffqALTXwevWDNywZEV0/wAg8vYxVdpkTyAz0vOIoGntEszxCWSMl5H3J5QXTue9a32sKhILBCL0XTyAwtoiu79h/W9n2HtGv/4XP35Q6pebwMkxfUeUNY3jbOztxdOYPXyzhUFYtHsxjRDt3RAntvu020PPDk/M60dUHoBdHRgLmqSWc0A1cnMJ8FdHPEoK0cUjzLW8ZnrfuBjahazDZaU+I91st4nJj8+4kV93lLlUSnt0f3O7fFhLc8i8crxI62k6jzfnui71kTndy91+gtxZPTMsNrg+M9X9VwbDQ4vmX3VRv+bb3rwePn/nYs9VhkrDA6Vdmu6JAj0IboMI2XifBOhF4g5A0PeUNTucuK5BSve68lekIfMq9555lJ4HZmRQCX0zLD+0zUqQOL1uPlFKqPMdtFYIWZRrCFkrwAparGJnpCifYs8hOdZDMAAAz3SURBVOQEN3XknPy91+INTXrvQ1VwGn1/osOnOar20Qt6c2ZSRu+rUwdWlS0tubbfGAvfGQoEVvGGF7wUfe5K9hifOtmXqy+7WE8++9JUNsAbHvI+j7t3bNW3n3t5KtuRs4/edusMQeUMyXm1R97woLeNg48+pTS5Xfn/vPc8+nqia9xFn29lAL3ejEbvOHjd0733zPtMeI/z9rvqMxrNlubwjuvQhwfJWAEBXpuIY3ffoFMH9ujY3Tecu4/FqGdbnU7X3gnq2Rd/NHUr+RkPL9vlffa8omsvO+EVWFdlT3Jm+0X/1ntvvPehah+99zwne5fT7LhOe4ucIdDopAIvu1j1GS09KzCa7Rr68CC//kBBK4Oss7zMVlW2y8uMYTi8k22dbI43NOVlPDZtkpaWpjuLe589r2bHy2J4GYKqz2id5XRWFzR7gUM0O+G9D1UTUrz3PJphimYNo6reL287XhbTE33PvMd5r6Xq9XnHMGcSUPSzk1MM3wcEVkDDvBOed5/kz3LEcOTM4KrinWSis9S8QM872XrDTS9OgrsXV80K9F6jN0vRO1l7gUPVcN5q0aCjan+898ILEqJtEbzAKDqbUfKH7ryaJu9iq2p4cbXo46rUCcIi2prt2TWGAoGeo0nfcFQFMqUbW0aHnL0hrOgwS9VQlzerLJqJ8B4XHVLzgo6q1xLdH6+eyhuuii4N5N1Xday8fffeCy+I9oYXvWPdt8LwWWnyTGAF9NzJA3t06sAenTyw59x93o81S/40x6tV8o6Bd19V7Us0OPL+ProosDfrNbrmXtXsuGjQkhNQRrvI1wkcvNfjvRbvPi8I8rbt3Ve1eLQXHEU76nsBnHesZ3m9vi4RWAED5BXOe1eDBFtvyMn8vf769K3k10N591VtN2eqfzTr5AU33snWK2iOzo6r2m8vIIx2MPfeG++zXBU4RD/33uPqTEBYzXvPqv7Wy8B5AZMXmPUtE4Vp1FgBI+al3L06hy5bR7S17ZwFe3OKab0C4CrR2pLo/kRrbLyAoM60+mixuBegerzPRJ39WW89xLXaFkQnIHi1YXX2sSoDt/q4erVlQ69BGjsCK2DGeCeT6GxGSZoz01JKmrMyVV9VQyWrdTljMnoi896zqr+Nzq7yApScAMwrXo/2tqrDCyijBejRz0QV73VH34tooB8N1OsUkHvHdRa60I8NQ4EAXN5woxTviu3xhl68RYE9Oy69eOq2j+qsrxetf/GKxaP1S942vGHEaG+rKl7Bt7ftaJAQrT+qeh+qGoxG3ouqz/1q3me+TgF59PVEh/2amCSBjSFjBaCW6BTs6FT2qnqa1dkSL6vSRHuDLnltAqLDWlFe5qbOLK3oMJ2XDfK2HV0U2NvHqvehjX5J0e9BVXYxelyj2cmcz8TQO533DRkrAI2IZmRysiU5dVN15KyPtmv/YW3fd0i79h+eut/LMHgBQXT2WVTuTLHoMJ2XDYoWzkdVZTujC02XnjVXJ2tU+rjmZLaG3um8b8hYAehUNFtSp+NzTndoT876aFXDg15TzmiNVnR9tSYyEV6Wx8swRrNB0YyMl82rem+9z1TpNe48TTTkLP233j5WHau2PlNjQ8YKwCB42YWqVgbeY3PWuMtR1a/K206dGq3VvKxDE5kI773dvWOr5symuojnZIOiPady1+b0sjfRrFNOPZSUd2xy6qm8faw6Vm19psaGwArAYNUZRvJqWrxga9NktuOmDcx69J6vqhja23cv2IqeREsPLVXp01BSnUDUC9YOTnptHVzRayv6WqJDnVXDyHUWZ45suwnePtJDa30EVgAGq05WxDsheMFWzomjzky6aL1R9CTaVuftaBfyHF7Ak8t7L3K6uUcfVzUE6X32qurxNrptT52gLNrhHdMIrADMhOjyLt7jvGxHW2s45mQ2mpBTJB3NvnkBj/d+564s4D1nNHDImZwh+cc1Zyg4qk5QRnZqYwisAMysnJNozmy2Kl42oY2i6zqi/aCi90WDLe/99mq76jyn1yOstKosZrSnl/daqtafjKiTcSI7tTFFAisz+6SZJTO7tMTzAUDfNXHSGWozSC/7Eq3PyRnqrPrb6HN22dU8eqwpFh+e7MDKzN4u6ZckcdQBIEN0NmNOxqIJXoASrc/J6eRfNUxauk6qCdFjXXof+xaUj1GJPlb3SPpNSV8p8FwAgBW8Qvi+dZyPrs3nqdOtfLWqjNNQFymu0wF/o3J7bWF9WRkrM9sr6XRK6Vih/QEArKOJ+q4cOcOipXsy1REdZosW4jeRDSo9FNi3yRBjtG5gZWZ/YmbHnX9uknSnpN+ObMjMbjOzBTNbWFxczN1vAJhZQygqzukHFZX7PuTUOXn35Q7ReoGZFwhFh0o9XdaVecY4NLluYJVS+mBK6ZrV/0j6jqQdko6Z2SlJV0r6mpn9eMXz3JtSmk8pzW/btq3kawAAZGji5Na3OqectQLbapTpBWule0n1rYXCGIvzN1xjlVL6hqTLzv73JLiaTyn9oMB+AQBa0kTdTbTOqa16qJy2FVXBTen9zqlVi6qz322sC1j69fUBizADwIwb48mtpLben5xCfo+3cHUdbRS6D3WiwVqKNQhNKW0nWwUAwzOEmq1cOXVJ0fcnZxtNyG0u27dhw6GwlNL6jypsfn4+LSwstL5dAADW0sbwV1tyM1Z4g5k9nlKajzyWoUAAACbG1OeJYKobrBUIAMAEw1/IxVAgAADAGuoMBZKxAgAAKITACgAAoBACKwAAgEIIrAAAAAohsAIAYODGuJjxUBFYAQAwcNHFjAnAmkdgBQDAwEX7b0UDMGwcndcBABi46GLGLLjdPBqEAgAArIEGoQAAAB0gsAIAACiEwAoAAKAQAisAAIBCCKwAAAAKIbACAAAohMAKAACgEAIrAACAQgisAAAACiGwAgAAKKSTJW3MbFHSU61vuJ5LJf2g653AeTgu/cWx6SeOS39xbPrJOy7vSClti/xxJ4HVEJjZQnRdILSH49JfHJt+4rj0F8emn3KPC0OBAAAAhRBYAQAAFEJgVe3erncALo5Lf3Fs+onj0l8cm37KOi7UWAEAABRCxgoAAKAQAqsJM5szs6+b2X+e/PffNLOvmtmJye2Pdb2Ps8jM3mZmD5jZX5jZE2b2sxyb7pnZ7Wb2TTM7bmZfNLMLOC7dMLN/b2bPmdnxFfdVHgszu8PMvmVmf2lmN3Sz1+NXcVx+d/Jb9r/N7D+Z2dtW/D+OS0u8Y7Pi/33SzJKZXbrivlrHhsDqDZ+Q9MSK/94n6U9TSjsl/enkv9G+35f0xymlvyNpl5aPEcemQ2Z2haTfkDSfUrpG0pykm8Vx6crnJH1o1X3usTCzd2n5WP29yd/8OzOba29XZ8rndP5x+aqka1JK75b0pKQ7JI5LBz6n84+NzOztkn5J0tMr7qt9bAisJJnZlZL2SPrMirtvkvRHk3//I0kfaXu/Zp2ZXSLp5yV9VpJSSq+mlP5aHJs+2CzpQjPbLOkiSd8Tx6UTKaX/Ien/rLq76ljcJOn+lNL/SymdlPQtST/Tyo7OGO+4pJSOpJTOTP7zUUlXTv6d49Kiiu+MJN0j6TclrSw+r31sCKyW/Wstv5mvr7jv8pTS9yVpcntZFzs2435S0qKkP5wM037GzC4Wx6ZTKaXTkn5Py1d135f0QkrpiDgufVJ1LK6Q9N0Vj3tmch/a9yuS/svk3zkuHTOzvZJOp5SOrfpftY/NzAdWZvZhSc+llB7vel9wns2S3ivpD1JK75H0shhe6tykXucmSTsk/YSki83s1m73CkHm3MfU8JaZ2Z2Szkj6wtm7nIdxXFpiZhdJulPSb3v/27lvzWMz84GVpPdL2mtmpyTdL+kDZnZQ0rNm9rckaXL7XHe7OLOekfRMSuno5L8f0HKgxbHp1gclnUwpLaaUXpP0ZUk/J45Ln1Qdi2ckvX3F467U8jAuWmJmH5P0YUm/nN7od8Rx6dbVWr5QPDaJBa6U9DUz+3Ft4NjMfGCVUrojpXRlSmm7lgvU/mtK6VZJD0n62ORhH5P0lY52cWallP5K0nfN7Kcmd/2ipD8Xx6ZrT0u61swuMjPT8nF5QhyXPqk6Fg9JutnM3mJmOyTtlPS/Oti/mWRmH5L0W5L2ppR+uOJ/cVw6lFL6RkrpspTS9kks8Iyk907OQbWPzebG93i4Dkj6kpn9Uy2fSP5Rx/szq35d0hfM7M2SviPp41q+IODYdCSldNTMHpD0NS0PZ3xdy52Kt4jj0joz+6KkX5B0qZk9I+luVfx+pZS+aWZf0vIFyhlJv5ZSWupkx0eu4rjcIektkr66fE2iR1NKv8pxaZd3bFJKn/Ueu5FjQ+d1AACAQmZ+KBAAAKAUAisAAIBCCKwAAAAKIbACAAAohMAKAACgEAIrAACAQgisAAAACiGwAgAAKOT/A1xeimuz8zxQAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Visualize the GC-count relationship\n",
    "fig = plt.figure(figsize=(10,5))\n",
    "ax = fig.add_subplot()\n",
    "\n",
    "temp_x = np.vstack([x[k] for k in x.keys() if 'unegs' not in k])\n",
    "temp_y = np.concatenate([y[k] for k in y.keys() if 'unegs' not in k])\n",
    "GC = np.sum(temp_x[:,(seqlen//2-seqlen1//2):(seqlen//2+seqlen1//2),:], axis=1)\n",
    "GC = GC[:,1]+GC[:,2]\n",
    "ax.scatter(GC/seqlen1*100, temp_y, s=4)\n",
    "coef = GCregress(temp_x[:,(seqlen//2-seqlen1//2):(seqlen//2+seqlen1//2),:], temp_y)    # limit GC counting to only the window where reads are counted\n",
    "temp_y -= coef*GC\n",
    "ax.scatter(GC/seqlen1*100, temp_y, s=4)\n",
    "ax.set_xlabel('GC content (%)')\n",
    "ax.set_ylabel('log accessibility')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Regress out GC content\n",
    "temp_x = np.vstack([x[k] for k in x.keys() if 'unegs' not in k])\n",
    "temp_y = np.concatenate([y[k] for k in y.keys() if 'unegs' not in k])\n",
    "coef = GCregress(temp_x[:,(seqlen//2-seqlen1//2):(seqlen//2+seqlen1//2),:], temp_y)    # limit GC counting to only the window where reads are counted\n",
    "\n",
    "\n",
    "cnt=0\n",
    "del x['test'], y['test']   # don't need these since test set comes from 'both' anyways\n",
    "for n in ['train','val']:\n",
    "    GC = np.sum(x[n][:,(seqlen//2-seqlen1//2):(seqlen//2+seqlen1//2),:], axis=1)\n",
    "    GC = GC[:,1]+GC[:,2]\n",
    "    y[n] -= coef*GC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-5.])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Do uneg count adjustment based on range of GC corrected values\n",
    "# print(np.min(y['train']))\n",
    "# print(np.unique(y['train_unegs']))\n",
    "# y['train_unegs'] -= 5\n",
    "# y['val_unegs'] -= 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "if save:\n",
    "    with h5py.File(datadir+'data'+ident+'_ref.h5','w') as f:\n",
    "        for key in x.keys():\n",
    "            f.create_dataset('x_'+key, data=x[key])\n",
    "            f.create_dataset('y_'+key, data=y[key])"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "0f07b599dba35ea3e3431fe483080f9a2b540e1c33071a57a1e300447d2107f4"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
